java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/utils/DateTimeUtils.java

```diff
@@ -47,7 +47,7 @@ public static long applyCalendarOffset(long milliseconds, Calendar calendar) {
     final TimeZone defaultTz = TimeZone.getDefault();
 
     if (tz != defaultTz) {
-      milliseconds -= tz.getOffset(milliseconds) - defaultTz.getOffset(milliseconds);
+      milliseconds += tz.getOffset(milliseconds) - defaultTz.getOffset(milliseconds);
```

Comment:
The problem is that I made this line of change in order to fix the test failures of this case: https://github.com/apache/arrow/blob/eaa1a1ea74e2b37020e0133f1460cb251e5f4974/java/flight/flight-sql-jdbc-core/src/test/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcTimeStampVectorAccessorTest.java#L292 and let me paste the code snippet of that case below:

```java
  @Test
  public void testShouldGetStringBeConsistentWithVarCharAccessorWithCalendar() throws Exception {
    // Ignore for TimeStamp vectors with TZ, as VarChar accessor won't consider their TZ
    Assume.assumeTrue(
        vector instanceof TimeStampNanoVector || vector instanceof TimeStampMicroVector ||
            vector instanceof TimeStampMilliVector || vector instanceof TimeStampSecVector);
    Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone(AMERICA_VANCOUVER));
    assertGetStringIsConsistentWithVarCharAccessor(calendar);
  }

  private void assertGetStringIsConsistentWithVarCharAccessor(Calendar calendar) throws Exception {
    try (VarCharVector varCharVector = new VarCharVector("",
        rootAllocatorTestRule.getRootAllocator())) {
      varCharVector.allocateNew(1);
      ArrowFlightJdbcVarCharVectorAccessor varCharVectorAccessor =
          new ArrowFlightJdbcVarCharVectorAccessor(varCharVector, () -> 0, (boolean wasNull) -> {
          });

      accessorIterator.iterate(vector, (accessor, currentRow) -> {
        final String string = accessor.getString();
        varCharVector.set(0, new Text(string));
        varCharVector.setValueCount(1);

        Timestamp timestampFromVarChar = varCharVectorAccessor.getTimestamp(calendar);
        Timestamp timestamp = accessor.getTimestamp(calendar);

        collector.checkThat(timestamp, is(timestampFromVarChar));
        collector.checkThat(accessor.wasNull(), is(false));
      });
    }
  }
```

Reply:
In the test above, the user-supplied timezone is `AMERICA_VANCOUVER`. When the input text to varCharVector is `2021-07-09 00:00:00.0`, varCharVectorAccessor.getTimestamp(calendar) returns `2021-07-09 07:00:00.0`. This does not make sense to me at all. @lidavidm

Reply:
To summarize, we have two bugs for now.
- conversion between time zones are incorrect.
- It assumes utc if time zone is not set in the timestamp vector. Same for varchar vector if it tries to interpret as timestamp values.

For the fix of HSBC implementation of getDate/getTimestamp with a calendar as input, I’d propose following:
- convert between timezones if timestamp vector has set timezone and user supplied calendar is not null.
- Throw unsupported operation exception if vector does not have timezone set and user has provided a timezone. The reason is that the returned timestamp object cannot carry timezone info.

IIUC, the timezone info in your derby example comes from the SimpleDateFormat but not the resultset itself.


WDYT? @lidavidm

Reply:
> varCharVectorAccessor.getTimestamp(calendar) returns 2021-07-09 07:00:00.0. This does not make sense to me at all.

That's [midnight in Vancouver expressed as GMT](https://www.wolframalpha.com/input?i=2021-07-09+07%3A00%3A00+GMT+in+Vancouver) so it's consistent with the other case described above. 

> conversion between time zones are incorrect.

It seems unintuitive, but at least consistent: it will take that date at midnight in the given time zone, and express it in GMT, which seems to match the docstring of java.sql.Date.

> Throw unsupported operation exception if vector does not have timezone set and user has provided a timezone. The reason is that the returned timestamp object cannot carry timezone info. Or return the timestamp as is to pretend it is in the user supplied timezone?

That may also work, but I suppose if the user is explicitly passing a Calendar, the current behavior is what they actually want? This is more a flaw of the original Java date/time APIs in that Date itself doesn't carry the timezone so the display value is weird.

Reply:
Right, at least these behaviors are consistent to each other. It is unintuitive since timestamp values are always returned in the GMT regardless of what calendar is supplied. I cannot say this is wrong because it seems to be aligned with the SQL `TIMESTAMP_WITH_TIMEZONE` type.

There was a [discussion](https://docs.google.com/document/d/1gNRww9mZJcHvUDCXklzjFEQGpefsuR_akCDfWsdE35Q) in the Hadoop community to work on the consistency of timestamp semantics across different SQL-on-Hadoop engines. It proposed a `TIMESTAMP_WITH_LOCAL_TIMEZONE` type in addition to the SQL `TIMESTAMP_WITHOUT_TIMEZONE` and `TIMESTAMP_WITH_TIMEZONE` types:
- SQL `TIMESTAMP_WITHOUT_TIMEZONE`: Always display the same wall clock time regardless of timezone.
- SQL `TIMESTAMP_WITH_TIMEZONE`: The timestamp value is stored in a specific timezone (can be non-UTC) and can be converted to any timezone for display.
- New `TIMESTAMP_WITH_LOCAL_TIMEZONE`: The timestamp value is stored in the UTC and always display it in the local timezone.

Most of the SQL engines in the aforementioned doc now support `TIMESTAMP_WITH_LOCAL_TIMEZONE` type and so do the open file formats like the Apache Parquet and Apache Orc. I made this PR to align the behavior with `TIMESTAMP_WITH_LOCAL_TIMEZONE` type. Nevertheless I am no an expert on JDBC so I am afraid that I may not know what is the expected behavior from the end users. 

Please feel free to close this PR if you think the current behavior is not an issue, or involve more people to the discussion if a fix is required. Thanks for the discussion in detail!

Reply:
Thanks for bearing with me. The mess of date/time APIs here is confusing...

For Arrow: `timestamp[s, TZ]` appears equivalent to `TIMESTAMP_WITH_LOCAL_TIMEZONE`. Arrow always stores the underlying value in UTC. `timestamp[s]` appears to be `TIMESTAMP_WITHOUT_TIMEZONE`. Arrow does not have an equivalent to `TIMESTAMP_WITH_TIMEZONE`. 

So the question here is still the semantics of `getDate(int, Calendar)`. The method looks ill-defined to me if `java.sql.Date` is not supposed to carry time or timezone information in the first place...The API docs say "This method uses the given calendar to construct an appropriate millisecond value for the date if the underlying database does not store timezone information."

I do see things like this: https://stackoverflow.com/questions/9202857/timezones-in-sql-date-vs-java-sql-date

> This means that when you work in UTC+1 and ask the database for a DATE a compliant implementation does exactly what you've observed: return a java.sql.Date with a milliseconds value that corresponds to the date in question at 00:00:00 UTC+1 independently of how the data got to the database in the first place.

So that also seems to match the existing behavior, by my reading.

I looked at Postgres's JDBC driver: 

https://github.com/pgjdbc/pgjdbc/blob/3c3e5b62349c42d9d3dfe1fe0102776d4de8c353/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L532

and

https://github.com/pgjdbc/pgjdbc/blob/3c3e5b62349c42d9d3dfe1fe0102776d4de8c353/pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java#L639-L653

> Truncate date part so in given time zone the date would be formatted as 00:00

I think that also matches with the current behavior, as unintuitive as it is.

Reply:
Thanks for the additional links! It is really a notorious topic :)